packages/next-on-pages/templates/_worker.js/utils/cache.ts

```diff
@@ -0,0 +1,176 @@
+import type { CacheInterface } from './cache-interface';
+import {
+	SUSPENSE_CACHE_URL,
+	getSuspenseCacheInterface,
+} from './cache-interface';
+
+/**
+ * Handles an internal request to the suspense cache.
+ *
+ * @param request Incoming request to handle.
+ * @returns Response to the request, or null if the request is not for the suspense cache.
+ */
+export async function handleSuspenseCacheRequest(request: Request) {
+	const baseUrl = `https://${SUSPENSE_CACHE_URL}/v1/suspense-cache/`;
+	if (!request.url.startsWith(baseUrl)) return null;
+
+	try {
+		const url = new URL(request.url);
+		const cache = await getSuspenseCacheInterface();
+
+		if (url.pathname === '/v1/suspense-cache/revalidate') {
+			// Update the revalidated timestamp for the tags in the tags manifest.
+			const tags = url.searchParams.get('tags')?.split(',') ?? [];
+			await cache.setTags(tags, { revalidatedAt: Date.now() });
+
+			return new Response(null, { status: 200 });
+		}
+
+		// Extract the cache key from the URL.
+		const cacheKeyId = url.pathname.replace('/v1/suspense-cache/', '');
+		if (!cacheKeyId.length) {
+			return new Response('Invalid cache key', { status: 400 });
+		}
+
+		const cacheKey = cache.buildCacheKey(cacheKeyId);
+
+		switch (request.method) {
+			case 'GET':
+				// Retrieve the value from the cache.
+				return handleRetrieveEntry(cache, cacheKey);
+			case 'POST':
+				// Update the value in the cache.
+				return handleUpdateEntry(cache, cacheKey, await request.text());
+			default:
+				return new Response(null, { status: 405 });
+		}
+	} catch (e) {
+		// eslint-disable-next-line no-console
+		console.error(e);
+		return new Response('Error handling cache request', { status: 500 });
+	}
+}
+
+/**
+ * Retrieves a value from the suspense cache.
+ *
+ * @param cache Interface for the suspense cache.
+ * @param cacheKey Key of the cached value to retrieve.
+ * @returns Response with the cached value.
+ */
+async function handleRetrieveEntry(cache: CacheInterface, cacheKey: string) {
+	// Get entry from the cache.
+	const entry = await cache.get(cacheKey);
+	if (!entry) return new Response(null, { status: 404 });
+
+	let data: CacheEntry;
+	try {
+		data = JSON.parse(entry) as CacheEntry;
+	} catch (e) {
+		return new Response('Failed to parse cache entry', { status: 400 });
+	}
+
+	// Load the tags manifest.
+	await cache.loadTagsManifest();
+
+	// Check if the cache entry is stale or fresh based on the tags.
+	const tags = getDerivedTags(data.value.data.tags ?? []);
+	const isStale = tags.some(tag => {
+		const tagEntry = cache.tagsManifest?.items?.[tag];
+		return (
+			tagEntry?.revalidatedAt && tagEntry?.revalidatedAt >= data.lastModified
+		);
+	});
+
```

Comment:
are you not missing the following part of the [fetch-cache](https://github.com/vercel/next.js/blob/canary/packages/next/src/server/lib/incremental-cache/fetch-cache.ts#L246) implementation? 
```typescript
    if (data?.value?.kind === 'FETCH') {
      const innerData = data.value.data
      const derivedTags = getDerivedTags(innerData.tags || [])

      if (
        derivedTags.some((tag) => {
          return this.revalidatedTags.includes(tag)
        })
      ) {
        data = undefined
      }
    }
```
If a tag got revalidated, it skips the cached entry to force a new fetch

Reply:
I actually didn't think about trying to implement that part because it would still run in the fetch cache, since this is just replacing the API endpoint.

After some testing though, it appears that the revalidated tags aren't properly tracked there, so I've just updated the code to track revalidated tags locally, so that server actions will be able to properly return the fresh data instead of requiring a refresh.

Thanks for getting me look at that part :)